### PR TITLE
Fix attack cursor during battle

### LINKED ISSUE #2497: Combat, attack cursor should adapt its direction according to the possible direction of the attack and in the largest possible area of the hexagon
In fheroes2, the attack cursor should adapt its direction according to the possible direction of the attack and in the largest possible area of the hexagon.

In fheroes2, the cursor must not display an attack direction that cannot be performed.
The cursor should automatically display the only possible attack direction and this in the maximum area of the hexagon where the attacked creature is located. (https://github.com/ihhub/fheroes2/issues/501)
In the example, the attack of the Dragons troop will not reach the Gargoyles troop contrary to what the direction of the attack cursor suggests.

In fheroes2:

Dragon attack direction fh2 01

Dragon attack direction fh2 02

In the original game, the cursor adapts its direction according to the possible direction of the attack and in the largest possible area of the hexagon.

In original game:

Dragon attack direction H2 01

Dragon attack direction H2 02

Dragon attack direction H2 03

Dragon attack direction H2 04

Dragon attack direction H2 05

Dragon attack direction H2 06

Rialdo attacks Dragontooth.

Here is the save file:
[Flying Hydras fh2 01.zip](https://github.com/ihhub/fheroes2/files/5745048/Flying.Hydras.fh2.01.zip)

**Edit 2020/12/29.**

Related issues: https://github.com/ihhub/fheroes2/issues/1628, https://github.com/ihhub/fheroes2/issues/1629.

### PR DESCRIPTION
**Pull request by ihhub**
close #2497

### CODE REVIEW

**Comment by oleg-derevenetz** [comment]
This doesn't fix the following issue:

https://user-images.githubusercontent.com/32623900/151677155-20824e6e-3782-434f-a723-0fffb2360396.mp4

Need to think how to fix it properly.

File: src/fheroes2/battle/battle_interface.cpp [hunk 1]
```diff
@@ -2129,18 +2129,106 @@ int Battle::Interface::GetBattleCursor( std::string & statusMsg ) const
...
+                            if ( availableAttackDirection.size() == 1 ) {
+                                currentDirection = *availableAttackDirection.begin();
+                            }
+                            else {
+                                // First seach clockwise.
+                                int clockWiseDirection = currentDirection;
+                                int antiClockWiseDirection = currentDirection;
+
+                                while ( true ) {
+                                    switch ( clockWiseDirection ) {
```
Comment (oleg-derevenetz) [inline src/fheroes2/battle/battle_interface.cpp:2161]: AFAIK there is the `direction_t & operator++( direction_t & d )` for that.

File: src/fheroes2/battle/battle_interface.cpp [hunk 2]
```diff
@@ -2129,18 +2129,106 @@ int Battle::Interface::GetBattleCursor( std::string & statusMsg ) const
...
+                                        assert( 0 );
+                                        break;
+                                    }
+
+                                    if ( availableAttackDirection.count( clockWiseDirection ) > 0 ) {
+                                        currentDirection = clockWiseDirection;
+                                        break;
+                                    }
+
+                                    switch ( antiClockWiseDirection ) {
```
Comment (oleg-derevenetz) [inline src/fheroes2/battle/battle_interface.cpp:2192]: AFAIK there is the `direction_t & operator--( direction_t & d )` for that.

File: src/fheroes2/battle/battle_interface.cpp (hunk 1 above)
Comment (ihhub) [reply src/fheroes2/battle/battle_interface.cpp:2161]: The problem is direction_t includes TOP and BOTTOM which aren't used here.

Comment (oleg-derevenetz) [reply src/fheroes2/battle/battle_interface.cpp:2161]: Hi @ihhub It seems that it doesn't:

```
namespace Battle
{
    class Unit;

    enum direction_t
    {
        UNKNOWN = 0x00,
        TOP_LEFT = 0x01,
        TOP_RIGHT = 0x02,
        RIGHT = 0x04,
        BOTTOM_RIGHT = 0x08,
        BOTTOM_LEFT = 0x10,
        LEFT = 0x20,
        CENTER = 0x40,
        RIGHT_SIDE = TOP_RIGHT | RIGHT | BOTTOM_RIGHT,
        LEFT_SIDE = TOP_LEFT | LEFT | BOTTOM_LEFT,
        AROUND = RIGHT_SIDE | LEFT_SIDE
    };
```

Comment (ihhub) [reply src/fheroes2/battle/battle_interface.cpp:2161]: Yes, you're absolutely right!

**Comment by ihhub** [comment]
> This doesn't fix the following issue:
> Free.Heroes.of.Might.and.Magic.II.version_.0.9.11.2022-01-29.23-48-46.mp4
> 
> Need to think how to fix it properly.

These changes just fix cases where we display a cross instead of attack direction. I'm not surprised about corner cases :)

**Review by oleg-derevenetz — APPROVED** [review]

**Pull request merged**